soh/soh/Enhancements/Fixes/DekuNutUpgradeFix.cpp

```diff
@@ -0,0 +1,43 @@
+#include <libultraship/bridge.h>
+#include "soh/Enhancements/game-interactor/GameInteractor_Hooks.h"
+#include "soh/ShipInit.hpp"
+#include "functions.h"
+#include "macros.h"
+#include "variables.h"
+#include "z64save.h"
+
+extern "C" PlayState* gPlayState;
+
+static constexpr int32_t CVAR_NUT_UPGRADE_FIX_DEFAULT = 0;
+#define CVAR_NUT_UPGRADE_FIX_NAME CVAR_ENHANCEMENT("DekuNutUpgradeFix")
+#define CVAR_NUT_UPGRADE_FIX_VALUE CVarGetInteger(CVAR_NUT_UPGRADE_FIX_NAME, CVAR_NUT_UPGRADE_FIX_DEFAULT)
+
+void DekuNutUpgradeFixAtForestStage(bool* should) {
+    // This check is needed because of an intentional fallthrough at the source
+    if (Player_GetMask(gPlayState) == PLAYER_MASK_SKULL) {
+        return;
+    }
+
+    s32 expectedNutUpgrades = (INV_CONTENT(ITEM_NUT) == ITEM_NUT ? 1 : 0) +
+                              (Flags_GetInfTable(INFTABLE_BOUGHT_NUT_UPGRADE) ? 1 : 0) +
+                              (Flags_GetItemGetInf(ITEMGETINF_OBTAINED_NUT_UPGRADE_FROM_STAGE) ? 1 : 0);
+    s32 actualNutUpgrades = CUR_UPG_VALUE(UPG_NUTS);
+
+    if (expectedNutUpgrades != actualNutUpgrades) {
+        Flags_UnsetItemGetInf(ITEMGETINF_OBTAINED_NUT_UPGRADE_FROM_STAGE);
+        *should = true;
+    }
+}
+
+void DekuNutUpgradeSetByPoachersSaw(bool* should) {
+    *should = false;
+}
+
+void RegisterDekuNutUpgradeFix() {
+    COND_VB_SHOULD(VB_POACHERS_SAW_SET_DEKU_NUT_UPGRADE_FLAG, CVAR_NUT_UPGRADE_FIX_VALUE || IS_RANDO,
+                   { DekuNutUpgradeSetByPoachersSaw(should); });
+    COND_VB_SHOULD(VB_DEKU_SCRUBS_REACT_TO_MASK_OF_TRUTH, CVAR_NUT_UPGRADE_FIX_VALUE || IS_RANDO,
```

Comment:
Based on my understanding of this fix and how rando handles item giving, this specific hook should not care about `IS_RANDO`, as this `VB_` wont even be triggered in a rando save to begin with.

I'd recommend removing the `|| IS_RANDO` so that it is clear the logic isn't applicable to rando. Also there should be an early return in the handler for `!IS_RANDO`

Reply:
Which hook are you referring to? It's not clear from this snippet.

I get removing `IS_RANDO` from the condition, but why is an early return necessary if these hooks won't be triggered in rando anyway?

Reply:
The one for mask of truth.
Based on my previous suggestion of having an additional call to `GameInteractor_Should(VB_DEKU_SCRUBS_REACT_TO_MASK_OF_TRUTH, true)` in the timesave code, that code is also used for rando saves, so we would need _something_ to prevent it from running when in a rando save.

Same goes for the case if you go the `onActorInit` route. Because the time saver can be on in rando saves, you'd need to explictly check agasint `IS_RANDO` to bail out.

Reply:
I've changed the condition to `&& !IS_RANDO` which should suffice.